### PR TITLE
dig internal cleanup

### PR DESCRIPTION
**Pull request by glibsm**
Wanted to PR this on it's own, to reduce the noise from the upcoming improvements to resolve some issues.

`dig.go` has grown rather large, so i split up some functionality out into their respective files. There is changes in functionality, purely moving things around.

I also recognize that dig might be a little difficult to work with internally, I will consider a more significant refactor (no externally visible changes) post-beta

### CODE REVIEW

File: dig/graph.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	}
+	return nil
+}
+
+// Reset the graph by removing all the registered nodes
+func (g *graph) Reset() {
+	g.Lock()
+	defer g.Unlock()
+
+	defaultGraph.nodes = make(map[interface{}]graphNode)
```
Comment (alsamylkin) [inline dig/graph.go]: inherited the bug..

File: dig/node.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	deps        []interface{}
+}
+
+// Call the function and return the result
+func (n *funcNode) value(g *graph) (reflect.Value, error) {
+	if n.cached {
+		return n.cachedValue, nil
+	}
+
+	cv := reflect.ValueOf(n.constructor)
```
Comment (alsamylkin) [inline dig/node.go]: Doesn't seem to be used for a long time, till line 111

File: dig/graph.go (hunk 1 above)
Comment (glibsm) [inline dig/graph.go]: oh man this is bad :(

File: dig/node.go (hunk 2 above)
Comment (glibsm) [inline dig/node.go]: yes, but any time you write reflect related function, you almost always need type and value, one or multiple times. So i got into a habit of doing that at the top, i don't see much harm in that

**Pull request merged**